index.ts

```diff
@@ -1,165 +1,257 @@
-    conversation.transcriptions.push({ startByte, endByte, text });
+interface TranscriptionEvent extends Event {
+  eventType: 'transcription';
+  data: {
+    buffer: Buffer;
+    transcription: string;
+    lastAudioByteEventTimestamp: number;
   }
 }
-const getTranscriptionSinceWatermark = (conversation: Conversation) => {
-  const { watermark, transcriptions } = conversation;
-  const sortedTranscriptions = [...transcriptions].sort((a, b) => b.endByte - a.endByte);
-  let transcriptionTexts = [];
-  const seenStartBytes = new Set();
+interface CutTranscriptionEvent extends Event {
+  eventType: 'cutTranscription';
+  data: {
+    buffer: Buffer;
+    transcription: string;
+    lastAudioByteEventTimestamp: number;
+  }
+}
+interface TalkEvent extends Event {
+  eventType: 'talk';
+  data: {
+    response: string;
+  }
+}
+interface EventLog {
+  events: Event[];
+}
+const eventlog: EventLog = {
+  events: []
+};
 
-  for (let transcription of sortedTranscriptions) {
-    if (transcription.startByte < watermark) {
-      break;
-    }
-    if (!seenStartBytes.has(transcription.startByte)) {
-      seenStartBytes.add(transcription.startByte);
-      transcriptionTexts.push(transcription.text);
+// EVENTLOG UTILITY FUNCTIONS
+// From the event log, get the transcription so far
+const getLastTranscriptionEvent = (): TranscriptionEvent => {
+  const transcriptionEvents = eventlog.events.filter(e => e.eventType === 'transcription');
+  return transcriptionEvents[transcriptionEvents.length - 1] as TranscriptionEvent;
+}
+const getCutTimestamp = (): number => {
+  const cutTranscriptionEvents = eventlog.events.filter(e => e.eventType === 'cutTranscription');
+  const lastCut = cutTranscriptionEvents.length > 0 ? cutTranscriptionEvents[cutTranscriptionEvents.length - 1].data.lastAudioByteEventTimestamp : eventlog.events[0].timestamp;
+  const responseReflexEvents = eventlog.events.filter(e => e.eventType === 'responseReflex');
+  const lastResponseReflex = responseReflexEvents.length > 0 ? responseReflexEvents[responseReflexEvents.length - 1].timestamp : eventlog.events[0].timestamp;
+  return Math.max(lastResponseReflex, lastCut);
+
+}
+const getTransciptionSoFar = (): string => {
+  const cutTranscriptionEvents = eventlog.events.filter(e => e.eventType === 'cutTranscription');
+  const lastTranscriptionEvent = getLastTranscriptionEvent();
+  const lastCutTranscriptionEvent = cutTranscriptionEvents[cutTranscriptionEvents.length - 1];
+  let transcription = cutTranscriptionEvents.map(e => e.data.transcription).join(' ');
+  if (!lastCutTranscriptionEvent || lastCutTranscriptionEvent.timestamp !== lastTranscriptionEvent.timestamp) {
+    transcription = transcription + (lastTranscriptionEvent?.data?.transcription || '')
+  }
+  return transcription
+}
+const getDialogue = (): string => {
+  const dialogueEvents = eventlog.events
+    .filter(e => e.eventType === 'responseReflex' || e.eventType === 'talk');
```

Comment:
if you're this deep, and you're pointing and laughing saying "omg linear growth", you would be correct
it's all over here, and I'm going to fix it